Pre commit hooks to check line endings and trailing whitespace

eriknw opened the pull request:
This is a follow up to #7861 to prevent regression by adding a pre-commit to check for CRLF line endings.

While here, I took the liberty to add a few sanity checks. These typically run very fast. They're typically not high-value, so also happy to remove 'em if anybody feels strongly against them.

Also, in the second commit I added a hook to remove trailing whitespace. This has been secretly nagging me since I started contributing to NetworkX.

dschult commented:
How much should we worry about these mixing up the blame? I know we can have `blame` ignore a PR/commit. But is it worth the trouble? And what kind of PRs are good candidates for ignoring?

eriknw commented:
> How much should we worry about these mixing up the blame? I know we can have blame ignore a PR/commit. But is it worth the trouble? And what kind of PRs are good candidates for ignoring?

I don't have a strong opinion either way, and I'm also okay with having lots of commits in `.git-blame-ignore-revs`. #7861 is probably a good candidate, and maybe this PR too, but it might be nice to break up the commits so that the changes to `.pre-commit-config.yaml` aren't ignored.

rossbar reviewed (approved):
As so often happens, I started expounding on larger themes rather than reviewing the actual changes :joy: . Re: these specific changes - I approve, thanks @eriknw ! Everything below is policy discussion, I'll add a bullet to the community meeting agenda!

---

> How much should we worry about these mixing up the blame? I know we can have blame ignore a PR/commit. But is it worth the trouble? And what kind of PRs are good candidates for ignoring?

I don't think it's a huge deal either way. Personally, I've never run into a situation (in networkx or any other project) where the blame has been so cluttered that I couldn't find what I wanted. In practice, I find `bisect` to be a much more useful tool, and I leave all revisions in for the bisecting process because there's no guarantee that linting didn't break something. IMO it'd be totally fine not to include any commits in `.git-blame-ignore-revs` because there's an `--ignore-rev` flag to allow users to do so themselves as needed - it's just a nice convenience.

Re: what to include in `.git-blame-ignore-revs`, I'd think any auto-generated linting-type changes would be good candidates. But again, *not* including commits in the ignore file isn't the end of the world (nor is having tons of commits in there).

> but it might be nice to break up the commits so that the changes to .pre-commit-config.yaml aren't ignored.

I generally agree with this - when it comes to linting, my vote would be:
 1) Have one commit with the changes to configs
 2) Squash all the auto-applied changes into another single commit
 3) Use the "rebase-and-merge" strategy to preserve the two separate commits
 4) After rebase, add the commit that contains all the squashed auto-changes to `.git-blame-ignore-revs`

eriknw commented:
Okay, I rewrote the commit history so one applies the changes and the other adds the pre-commit hooks. Since I put forth the effort to do this, I went ahead and added these linting changes and those from #7861 to `.git-blame-ignore-revs`.

I rarely use `git reset --hard` and `git push --force`, so this was kinda fun.

dschult reviewed (commented):
Questions about the new pre-commit hooks:
- does this stop us from e.g. adding large files? How do we override the checks if needed?
- does this make it harder for contributors (like me:) to run pre-commit? That is, do they need to install the pre-commit-hooks repo manually? I would hope it is part of installing pre-commit. When and how would they need to/know to update the pre-commit?
- I'm not sure what counts as a submodule. Is this the repo within a repo kind of submodule?
- I'm still unclear about when pre-commit gets run. For sure it is run when I commit locally and have it installed. Is it also part of CI? Is it part of the "Squash and Merge" process when we merge a PR?

I'm definitely in favor of the mixed-line-ending and trailing-whitespace checks.
Thanks!

eriknw commented:
Great questions!

> * does this stop us from e.g. adding large files?

Yes, it prevents us from _accidentally_ adding large files (default=500kB).

> * How do we override the checks if needed?

`git commit --no-verify`

> * does this make it harder for contributors (like me:) to run pre-commit? That is, do they need to install the pre-commit-hooks repo manually? I would hope it is part of installing pre-commit. When and how would they need to/know to update the pre-commit?

Nope, it's not harder to use. pre-commit installs hooks automatically, so if you've already done `pre-commit install`, nothing more needs to be done.

> * I'm not sure what counts as a submodule. Is this the repo within a repo kind of submodule?

Yes, it means a "repo within a repo" kind of git submodule. Let's hope networkx never needs this.

> * I'm still unclear about when pre-commit gets run. For sure it is run when I commit locally and have it installed. Is it also part of CI? Is it part of the "Squash and Merge" process when we merge a PR?

Yes, it gets run locally when you commit if you have done `pre-commit install`, which is part of the contributing guide. It also runs in CI for PRs and commits, including "squash and merge": https://github.com/networkx/networkx/blob/main/.github/workflows/lint.yml

rossbar reviewed (approved):
Thanks @eriknw ! I'm +1 on all the added hooks!

dschult reviewed (approved):
I approve these changes too...  We should keep an eye on the ignore hashes and update as needed. And if linters get too noisy we can take them back out again. :)

eriknw commented:
Dope! I forgot to remind the person who merges this to _not_ squash the commits as per @rossbar's discussion. So, we can remove an entry from `.git-blame-ignore-revs`

dschult commented:
Ack!  that was me :(   Sorry....
